ui/components/mission/MissionPayRedeem.tsx

```diff
@@ -0,0 +1,572 @@
+            <h3 className="opacity-60 text-sm">Current Supply</h3>
+            <p>
+              {Math.floor(token?.tokenSupply.toString() / 1e18)} $
+              {token?.tokenSymbol}
+            </p>
+          </div>
+
+          <MissionTokenExchangeRates
+            currentStage={currentStage}
+            tokenSymbol={token?.tokenSymbol}
+          />
+        </div>
+        {tokenBalance > 0 && stage === 4 && (
+          <div
+            id="mission-redeem-container"
+            className="p-2 bg-darkest-cool rounded-2xl flex flex-col gap-4"
+          >
+            <div>
+              <h3 className="opacity-60 text-sm">Your Balance</h3>
+              <p className="text-2xl">{`${tokenBalance.toFixed(2)} $${
+                token?.tokenSymbol
+              }`}</p>
+            </div>
+
+            <PrivyWeb3Button
+              className="w-full rounded-full py-2"
+              label="Redeem"
+              action={redeem}
+              noPadding
+            />
+          </div>
+        )}
+      </div>
+    </div>
+  )
+}
+
+export type MissionPayRedeemProps = {
+  selectedChain: any
+  mission: any
+  token: any
+  fundingGoal: number
+  subgraphData: any
+  teamNFT: any
+  stage: any
+  ruleset: any
+  onlyModal?: boolean
+  modalEnabled?: boolean
+  setModalEnabled?: (enabled: boolean) => void
+  primaryTerminalAddress: string
+  forwardClient?: any
+}
+
+export default function MissionPayRedeem({
+  selectedChain,
+  mission,
+  token,
+  fundingGoal,
+  subgraphData,
+  teamNFT,
+  stage,
+  ruleset,
+  onlyModal = false,
+  modalEnabled = false,
+  setModalEnabled,
+  primaryTerminalAddress,
+  forwardClient,
+}: MissionPayRedeemProps) {
+  const chainSlug = getChainSlug(selectedChain)
+  const router = useRouter()
+  const [missionPayModalEnabled, setMissionPayModalEnabled] = useState(false)
+  const account = useActiveAccount()
+  const address = account?.address
+  const [input, setInput] = useState(0)
+  const [output, setOutput] = useState(0)
+  const [message, setMessage] = useState('')
+
+  const nativeBalance = useNativeBalance()
+
+  const { fundWallet } = useFundWallet()
+  const [agreedToCondition, setAgreedToCondition] = useState(false)
+
+  const currentStage = useMissionFundingStage(mission?.id)
+
+  const primaryTerminalContract = useContract({
+    address: primaryTerminalAddress,
+    chain: selectedChain,
+    abi: JBMultiTerminalABI as any,
+    forwardClient,
+  })
+
+  const tokenBalance = useWatchTokenBalance(selectedChain, token?.tokenAddress)
+
+  const getQuote = useCallback(async () => {
+    if (!address) return
+    if (!primaryTerminalContract) {
+      console.error('Primary terminal contract not initialized')
+      return
+    }
+
+    try {
+      const transaction = prepareContractCall({
+        contract: primaryTerminalContract,
+        method: 'pay' as string,
+        params: [
+          mission?.projectId,
+          '0x000000000000000000000000000000000000EEEe',
```

Comment:
can we pull this out to a named constant? not clear what this is without context